0101-nix formatting.md

```diff
@@ -0,0 +1,635 @@
+[
+  "--some-flag" "some-value"
+]
+```
+
+is formatted as
+
+```
+[
+  "--some-flag"
+  "some-value"
+]
+```
+
+workaround:
+
+```
+lib.cli.toGNUCommandLine {} {
+  some-flag = "some-value";
+}
+```
+
+### Documentation
+
+Update [Section 20.1 (Coding conventiones → Syntax)](https://nixos.org/manual/nixpkgs/stable/#sec-syntax) of the Nixpkgs manual. All formatting-specific guidance is removed and replaced with instructions on how to automatically format Nixpkgs instead.
+
+### Automatically generated files
+
+TODO discuss that https://github.com/NixOS/rfcs/pull/101#discussion_r690272696
+- exclude option in tooling
+- 2nix apply fomatting style
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Having a commit that changes the formatting, can make git blame harder to use. It will need `--ignore-rev` flag.
+- Every formatter will have bugs.
+
+## Alternatives
+
+[alternatives]: #alternatives
+
+- Keep the status quo of not having an official formatter. The danger is that this creates discord within the nix community. On top of fragmented the community it can generate lengthy discussions (which do not advance the eco-system).
+
+### Incremental migration
+
+A more gradual approach, where only changed files get formatted for some period of time to make a more graceful transition has been rejected for practical concerns. Among other things, such an approach would not improve the amount of merge conflicts, and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+[unresolved]: #unresolved-questions
+
+- Are there situation where automated formatting is worse than manual formatting? Do we want to have exceptions to automated formatting?
+
+## Future work
+
+[future]: #future-work
+
+- TODO General style guidelines beyond AST reformatting
+
+## Appendix A: Formatting style guidelines
+
+**Nomenclature:**
+Brackets: `[]`
+Braces: `{}`
+Parentheses: `()`
+
+*These are not strict rules but guidelines to help create more consistent formatting rules.
+Nevertheless, deviations should be documented and explained.*
+
+- When deciding between two *equally good* options, currently prevalent formatting style in `nixpkgs` should be followed.
+- On the top level (indentation zero), special rules may apply.
+- Two spaces are used for each indentation level, there is no vertical alignment
+- Consistency across language features is important. Syntax that behaves similarly should be formatted similarly.
+- The Nix formatting style is space heavy. Where there can be a space, there should be a space.
+  - Brackets and braces are generally written with a space on the inside, like `[ `, ` ]`, `{ ` and ` }`.
+  - `[]` is written as `[ ]`, same for `{ }`.
+  - Exception: Parentheses are written *without* a space on the inside
+- Expressions are either written on one line, or maximally spread out across lines, with no in-between (e.g. grouping).
+  - Multi-line statements increment the indentation level in the statement's "body".
```

Comment:
The specific rules below carve out one exception to this bullet point for top-level function bodies. I'd like to propose a few more (not limited to the top level):

```nix
assert ...;
big multi-line body {
  goes = here;
}
```

```nix
with ...;
big multi-line body {
  goes = here;
}
```

```nix
let
  x = ...;
in
big multi-line body {
  using = x;
}
```

These exceptions are motivated by my expectation that adding/removing assertions, factoring/inlining a `with`, and eliminating/inlining common subexpressions with `let` are all common code transformations that typically change only a few lines in the body (or zero, in the case of `assert`), and diffs would be less focused if performing these transformations required reindenting the entire body.

If this style guide doesn't forbid indentation in these bodies, I would at least want it not to be required.

Reply:
please, unneeded indentation is always so unpleasing
like if the last `}` ends on indentation level 1 instead of 0
alejandra does this